src/cargo/util/config.rs

```diff
@@ -154,7 +154,7 @@ impl Config {
     /// Get the path to the `cargo` executable
     pub fn cargo_exe(&self) -> CargoResult<&Path> {
         self.cargo_exe.get_or_try_init(||
-            env::current_exe().and_then(|path| path.canonicalize())
+            env::var(::CARGO_ENV).map(PathBuf::from).or_else(|_| env::current_exe().and_then(|path| path.canonicalize()))
```

Comment:
I didn't canonicalise in both cases on purpose, i.e.:

```rust
            env::var(::CARGO_ENV).map(PathBuf::from).or_else(|_| env::current_exe()).and_then(|path| path.canonicalize())
```

I figure if the user is supplying their own, we can trust them to select an appropriate incantation. (And if this is one we've propagated ourselves, we'll have canonicalised it here first.)